zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+
+    /// Return the hash of the next checkpoint we want to verify.
+    ///
+    /// If verification has not started yet, returns the hash of block zero
+    /// (the genesis block). If verification has finished, returns None.
+    fn get_next_checkpoint_hash(&self) -> Option<BlockHeaderHash> {
+        match self.get_next_checkpoint_height() {
+            // Every checkpoint height must have a hash
+            Some(height) => Some(*self.checkpoint_list.get(&height).unwrap()),
+            None => None,
+        }
+    }
+
+    /// Return the height of the next checkpoint higher than `after_block_height`
+    /// in the checkpoint list. Ignores the current checkpoint range.
+    ///
+    /// If `after_block_height` is None, assume verification has not started yet,
+    /// and return zero (the genesis block).
+    ///
+    /// Returns None if there are no checkpoints. Also returns None if
+    /// `after_block_height` is greater than or equal to the maximum height.
+    fn find_descendant_checkpoint_height(
+        &self,
+        after_block_height: Option<BlockHeight>,
+    ) -> Option<BlockHeight> {
+        match after_block_height {
+            None => self.checkpoint_list.keys().cloned().next(),
+            Some(height) => self
+                .checkpoint_list
+                .range((Excluded(height), Unbounded))
+                .next()
+                .map(|(height, _)| *height),
+        }
+    }
+
+    /// Increase the current checkpoint height to `verified_block_height`,
+    /// if `verified_block_height` is the next checkpoint height.
+    fn update_current_checkpoint_height(&mut self, verified_block_height: BlockHeight) {
+        let previous_height = self.get_previous_checkpoint_height();
+        // Ignore blocks that are below the previous checkpoint.
+        //
+        // We ignore out-of-order verification, such as:
+        //  - the height is less than the previous checkpoint height, or
+        //  - the previous checkpoint height is the maximum height (checkpoint verifies are finished),
+        // because futures might not resolve in height order.
+        if let Some(previous) = previous_height {
+            if verified_block_height <= previous {
+                return;
+            }
+        }
+
+        // Ignore updates if the checkpoint list is empty, or verification has finished.
+        if let Some(next_height) = self.get_next_checkpoint_height() {
+            if verified_block_height != next_height {
+                return;
+            }
+        }
+
+        // Set the new range
+        if let Some(new_checkpoint) =
+            self.find_descendant_checkpoint_height(Some(verified_block_height))
+        {
+            // Increment the range
+            self.current_checkpoint_range =
+                Some((Excluded(verified_block_height), Included(new_checkpoint)));
+        } else {
+            // Verification has finished
+            self.current_checkpoint_range = None;
+        }
```

Comment:
TIOLI: you can probably rewrite this with `Option`'s combinators

```rust
self.current_checkpoint_range = self.find_decendent_checkpoint_height(Some(verified_block_height))
    .map(|new_checkpoint| Some((Excluded(verified_block_height), Included(new_checkpoint)));
```

Reply:
Fixed in 018bff414adc012faecd6b77b953f04ebe566930.